File: 347-top-k-frequent-elements/top-k-frequent-elements.cpp
class Solution {
public:
    vector<int> topKFrequent(vector<int>& nums, int k) {
        unordered_map<int, int> mp;
        for (auto& num : nums)
            mp[num]++;

        vector<vector<int>> bucket(nums.size() + 1);
        for (auto& i : mp)
            bucket[i.second].push_back(i.first);

        vector<int> ans;
        for (int i = bucket.size() - 1; i >= 0; i--) {
            for (int j = bucket[i].size() - 1; j >= 0; j--) {
                ans.push_back(bucket[i][j]);
                if (ans.size() == k)
                    return ans;
            }
        }
        return ans;
    }
};

File: 347-top-k-frequent-elements/top-k-frequent-elements.java
class Solution {
    public int[] topKFrequent(int[] nums, int k) {
        Map<Integer, Integer> map = new HashMap<>();
        for (int num : nums) {
            map.put(num, map.getOrDefault(num, 0) + 1);
        }
        int maxi = 0;
        for (Integer val : map.values())
            maxi = Math.max(maxi, val);
        List<List<Integer>> bucket = new ArrayList<>();
        for (int i = 0; i <= maxi; i++)
            bucket.add(new ArrayList<>());
        for (Integer key : map.keySet()) {
            int freq = map.get(key);
            bucket.get(freq).add(key);
        }
        List<Integer> res = new ArrayList<>();
        for (int i = maxi; i >= 0; i--) {
            for (Integer num : bucket.get(i)) {
                res.add(num);
                k--;
                if (k == 0)
                    break;
            }
            if (k == 0)
                break;
        }
        return res.stream().mapToInt(Integer::intValue).toArray();
    }
}